cpp/test/mdarray.cu

```diff
@@ -417,4 +417,424 @@ TEST(MDArray, FuncArg)
+    constexpr int n_cols = 5;
+    int data_row_major[] = {
+      1,
+      2,
+      3,
+      4,
+      5, /* X  X  X */
+      6,
+      7,
+      8,
+      9,
+      10 /* X  X  X */
+    };
+    // manually aligning the above, using -1 as filler
+    static constexpr int X = -1;
+    int data_padded[]      = {1, 2, 3, 4, 5, X, X, X, 6, 7, 8, 9, 10, X, X, X};
+
+    using extents_type       = stdex::extents<stdex::dynamic_extent, stdex::dynamic_extent>;
+    using padded_mdspan      = stdex::mdspan<int, extents_type, SimplePaddedLayout2D>;
+    using padded_layout_type = typename SimplePaddedLayout2D::template mapping<extents_type>;
+    using row_major_mdspan   = stdex::mdspan<int, extents_type, stdex::layout_right>;
+
+    auto padded = padded_mdspan(data_padded, padded_layout_type(extents_type(n_rows, n_cols), 8));
+    auto row_major = row_major_mdspan(data_row_major, n_rows, n_cols);
+
+    int failures = 0;
+    for (int irow = 0; irow < n_rows; ++irow) {
+      for (int icol = 0; icol < n_cols; ++icol) {
+        if (padded(irow, icol) != row_major(irow, icol)) { ++failures; }
+      }
+    }
+    ASSERT_EQ(failures, 0);
+  }
+}
+
+TEST(MDSpan, PaddingCustomLayout) { test_mdspan_padding_custom_layout(); }
+
+void test_mdspan_padding_layout_stride()
+{
+  {
+    // 5x2 example,
+    constexpr int n_rows    = 2;
+    constexpr int n_cols    = 5;
+    constexpr int alignment = 8;
+
+    int data_row_major[] = {
+      1,
+      2,
+      3,
+      4,
+      5, /* X  X  X */
+      6,
+      7,
+      8,
+      9,
+      10 /* X  X  X */
+    };
+    // manually aligning the above, using -1 as filler
+    static constexpr int X = -1;
+    int data_padded[]      = {1, 2, 3, 4, 5, X, X, X, 6, 7, 8, 9, 10, X, X, X};
+
+    using extents_type     = stdex::extents<stdex::dynamic_extent, stdex::dynamic_extent>;
+    using padded_mdspan    = stdex::mdspan<int, extents_type, stdex::layout_stride>;
+    using row_major_mdspan = stdex::mdspan<int, extents_type, stdex::layout_right>;
+    layout_padded_right<extents_type, alignment> layout{extents_type{n_rows, n_cols}};
+
+    auto padded    = padded_mdspan(data_padded, layout);
+    auto row_major = row_major_mdspan(data_row_major, n_rows, n_cols);
+
+    int failures = 0;
+    for (int irow = 0; irow < n_rows; ++irow) {
+      for (int icol = 0; icol < n_cols; ++icol) {
+        if (padded(irow, icol) != row_major(irow, icol)) { ++failures; }
+      }
+    }
+    ASSERT_EQ(failures, 0);
+  }
+}
+
+TEST(MDSpan, PaddingLayoutStride) { test_mdspan_padding_layout_stride(); }
+
+void test_mdarray_padding()
+{
+  using extents_type = stdex::extents<dynamic_extent, dynamic_extent>;
+  auto s             = rmm::cuda_stream_default;
+
+  {
+    constexpr int rows      = 6;
+    constexpr int cols      = 7;
+    constexpr int alignment = 5;
+
+    /**
+     * padded device array
+     */
+    using padded_mdarray_type = device_mdarray<float, extents_type, stdex::layout_stride>;
```

Comment:
I think, it's worth a try to implement the padded layouts in a way to plug them in the [`LayoutPolicy`](https://github.com/kokkos/mdspan/blob/d861a3c6eb8cf5da2baa877d7eac114e7fd42453/include/experimental/__p0009_bits/mdspan.hpp#L59) template parameter directly rather than using the runtime polymorphism.
If that is not possible with the current implementation of the mdspan, maybe it would make sense to communicate the problem to the mdarray designers and do the necessary changes upstream, while it's not too late? Otherwise, what's the point of the `LayoutPolicy` being a template parameter? :)